File: Source/Remote Control Server/Screenshot.vb
Imports System.Drawing
Imports System.Threading
Imports System.Windows.Threading

Module Screenshot

    Public screenIndex As Integer = 0 'For multiple monitors
    Public isSendingBitmap As Boolean = False

    Public averageColor As Color = Color.White
    Private lastAverageColor As Color = Color.White

    Private colorDispatcherTimer As DispatcherTimer
    Private colorDispatcherActive As Boolean = False
    Private colorUpdateInterval As Integer = 2 'Update average color each x seconds

    Public Function getScreenShot(ByVal fullscreen As Boolean, ByVal index As Integer) As Bitmap
        Dim screenGrab As Bitmap
        Try
            'Get current screen
            Dim locations As Point() = getScreenBounds(index)
            Dim startLocation As Point = locations(0)
            Dim endLocation As Point = locations(1)
            Dim screenSize As Size = New Size(endLocation.X - startLocation.X, endLocation.Y - startLocation.Y)

            screenGrab = New Bitmap(screenSize.Width, screenSize.Height)
            Dim g As Graphics = Graphics.FromImage(screenGrab)
            g.CopyFromScreen(startLocation, New Point(0, 0), screenSize)

            'Scale image down
            Dim gr_dest As Graphics
            Dim bm_dest As Bitmap
            If fullscreen Then
                bm_dest = New Bitmap(CInt(screenGrab.Width * Settings.screenScaleFull), CInt(screenGrab.Height * Settings.screenScaleFull))
            Else
                bm_dest = New Bitmap(CInt(screenGrab.Width * Settings.screenScale), CInt(screenGrab.Height * Settings.screenScale))
            End If
            gr_dest = Graphics.FromImage(bm_dest)
            gr_dest.DrawImage(screenGrab, 0, 0, bm_dest.Width + 1, bm_dest.Height + 1)
            screenGrab = bm_dest

            'Cut off black borders (Powerpoint)
            removeBlackBorders(screenGrab)

        Catch ex As Exception
            screenGrab = New Bitmap(My.Resources.ic_action_warning)
            Logger.add("Can't get screenshot:" & vbNewLine & ex.ToString)
        End Try
        Return screenGrab
    End Function

    Public Sub removeBlackBorders(ByRef bitmap As Bitmap)
        Dim x As Integer = 0
        Dim y As Integer = bitmap.Height / 2
        Dim color_left As Color
        Dim color_right As Color
        Dim isBlack As Boolean = True
        Try
            Do While isBlack
                color_left = bitmap.GetPixel(x, y)
                color_right = bitmap.GetPixel(bitmap.Width - x - 1, bitmap.Height - y - 1)

                If color_left.GetBrightness = 0 And color_right.GetBrightness = 0 Then
                    x += 5
                Else
                    isBlack = False
                End If
            Loop

            Dim CropRect As New Rectangle(x, y, bitmap.Width - x - x, bitmap.Height - y - y)
            Dim CropImage = New Bitmap(CropRect.Width, CropRect.Height)
            Using grp = Graphics.FromImage(CropImage)
                grp.DrawImage(bitmap, New Rectangle(0, 0, CropRect.Width, CropRect.Height), CropRect, GraphicsUnit.Pixel)
            End Using
            bitmap = CropImage
        Catch ex As Exception
            'Logger.add("Can't crop bitmap:" & vbNewLine & ex.ToString)
        End Try
    End Sub

    Public Function getScreenBounds(ByVal index As Integer) As Point()
        Dim screens As Forms.Screen() = Forms.Screen.AllScreens
        Dim startLocation As New Point(0, 0)
        Dim endLocation As New Point(0, 0)

        If index >= 0 Then
            Dim current_screen As Forms.Screen
            If (index < screens.Count) Then
                current_screen = screens.ElementAt(index)
            Else
                current_screen = screens.ElementAt(0)
                screenIndex = 0
            End If

            startLocation.X = current_screen.Bounds.X
            startLocation.Y = current_screen.Bounds.Y
            endLocation.X = current_screen.Bounds.X + current_screen.Bounds.Width
            endLocation.Y = current_screen.Bounds.Y + current_screen.Bounds.Height
        Else
            For Each tempScreen As Forms.Screen In screens
                If tempScreen.Bounds.X < startLocation.X Then
                    startLocation.X = tempScreen.Bounds.X
                End If

                If tempScreen.Bounds.Y < startLocation.Y Then
                    startLocation.Y = tempScreen.Bounds.Y
                End If

                If tempScreen.Bounds.X + tempScreen.Bounds.Width > endLocation.X Then
                    endLocation.X = tempScreen.Bounds.X + tempScreen.Bounds.Width
                End If

                If tempScreen.Bounds.Y + tempScreen.Bounds.Height > endLocation.Y Then
                    endLocation.Y = tempScreen.Bounds.Y + tempScreen.Bounds.Height
                End If
            Next
        End If
        Dim locations As Point() = New Point() {startLocation, endLocation}
        Return locations
    End Function

    Public Sub sendBitmap(ByVal bmp As Bitmap, ByVal quality As Integer)
        Try
            isSendingBitmap = True
            Dim command As New Command
            command.source = Network.getServerIp()
            command.destination = Remote.lastCommand.source
            command.priority = RemoteControlServer.Command.PRIORITY_MEDIUM
            command.data = Converter.bitmapToByte(bmp, quality)
            command.send()
        Catch ex As Exception
            Logger.add("Can't send bitmap:" & vbNewLine & ex.ToString)
        End Try
        isSendingBitmap = False
    End Sub

    Public Sub sendScreenshot(ByVal full As Boolean)
        If Not isSendingBitmap Then
            If full Then
                sendBitmap(getScreenShot(True, screenIndex), Settings.screenQualityFull)
            Else
                sendBitmap(getScreenShot(False, screenIndex), Settings.screenQuality)
            End If
        End If
    End Sub

    'Shows the next screen if multiple screens available
    Public Sub toogleScreen()
        Dim new_index As Integer = screenIndex + 1
        Dim screens As Forms.Screen() = Forms.Screen.AllScreens
        If (new_index > screens.Count - 1) Then
            'Include all screens
            new_index = -1
        End If
        screenIndex = new_index
    End Sub

    Public Sub startUpdateColorTimer()
        If Not colorDispatcherActive Then
            colorDispatcherTimer = New DispatcherTimer
            AddHandler colorDispatcherTimer.Tick, AddressOf updateAverageColorTimerTick
            colorDispatcherTimer.Interval = New TimeSpan(0, 0, colorUpdateInterval)
            colorDispatcherTimer.Start()
            colorDispatcherActive = True
        End If
    End Sub

    Private Sub updateAverageColorTimerTick(ByVal sender As Object, ByVal e As EventArgs)
        Dim updateColorThread = New Thread(AddressOf updateAverageColor)
        updateColorThread.IsBackground = True
        updateColorThread.Start()
    End Sub

    Public Sub updateAverageColor()
        If Settings.updateAmbientColor Then
            averageColor = Screenshot.getApproximateAverageColor(Screenshot.getScreenShot(False, Screenshot.screenIndex))
            If Settings.serialCommands And lastAverageColor <> averageColor Then
                'Send fade to color command to arduino
                Dim message As String = "<22" & _
                    Converter.byteToAsciiNumber(averageColor.R) & _
                    Converter.byteToAsciiNumber(averageColor.G) & _
                    Converter.byteToAsciiNumber(averageColor.B) & _
                    "9" & _
                    ">"
                Serial.sendMessage(message)
                lastAverageColor = averageColor
            End If
        End If
    End Sub

    Public Function getAverageColor(ByVal bmp As Bitmap) As Color
        Dim totalR As Integer = 0
        Dim totalG As Integer = 0
        Dim totalB As Integer = 0
        Dim count As Integer = 1

        For x As Integer = 0 To bmp.Width - 1
            For y As Integer = 0 To bmp.Height - 1
                Dim pixel As Color = bmp.GetPixel(x, y)
                If pixel.GetSaturation > 0.6 And pixel.GetBrightness > 0.2 Then
                    totalR += pixel.R
                    totalG += pixel.G
                    totalB += pixel.B
                    count += 1
                End If
            Next
        Next

        'Dim totalPixels As Integer = bmp.Height * bmp.Width
        Dim averageR As Integer = totalR / count
        Dim averageG As Integer = totalG / count
        Dim averageB As Integer = totalB / count

        Return Color.FromArgb(averageR, averageG, averageB)
    End Function

    Public Function getApproximateAverageColor(ByVal bmp As Bitmap) As Color
        Dim totalR As Integer = 0
        Dim totalG As Integer = 0
        Dim totalB As Integer = 0
        Dim count As Integer = 1

        For x As Integer = 0 To bmp.Width - 1
            For y As Integer = 0 To bmp.Height - 1
                Dim pixel As Color = bmp.GetPixel(x, y)
                If pixel.GetSaturation > 0.0 And pixel.GetBrightness > 0.0 Then
                    totalR += pixel.R
                    totalG += pixel.G
                    totalB += pixel.B
                    count += 1
                End If
            Next
        Next

        'Dim totalPixels As Integer = bmp.Height * bmp.Width
        Dim averageR As Integer = totalR / count
        Dim averageG As Integer = totalG / count
        Dim averageB As Integer = totalB / count

        Dim roundUp As Byte = 200
        Dim roundDown As Byte = 100

        If averageR > roundUp Then
            averageR = 255
        ElseIf averageR < roundDown Then
            averageR = 0
        End If

        If averageG > roundUp Then
            averageG = 255
        ElseIf averageG < roundDown Then
            averageG = 0
        End If

        If averageB > roundUp Then
            averageB = 255
        ElseIf averageB < roundDown Then
            averageB = 0
        End If

        Return Color.FromArgb(averageR, averageG, averageB)
    End Function

    Public Function getRealAverageColor(ByVal bmp As Bitmap) As Color
        Dim totalR As Integer = 0
        Dim totalG As Integer = 0
        Dim totalB As Integer = 0

        For x As Integer = 0 To bmp.Width - 1
            For y As Integer = 0 To bmp.Height - 1
                Dim pixel As Color = bmp.GetPixel(x, y)
                totalR += pixel.R
                totalG += pixel.G
                totalB += pixel.B
            Next
        Next

        Dim totalPixels As Integer = bmp.Height * bmp.Width
        Dim averageR As Integer = totalR \ totalPixels
        Dim averageg As Integer = totalG \ totalPixels
        Dim averageb As Integer = totalB \ totalPixels

        Return Color.FromArgb(averageR, averageg, averageb)
    End Function

End Module

File: Source/MacOS/RemoteControlServer/screenshot.cpp
#include "apiv3.h"
#include "converter.h"
#include "network.h"
#include "remote.h"
#include "screenshot.h"
#include "serial.h"
#include "settings.h"

#include <iostream>
#include <thread>
#include <chrono>

#include <QApplication>
#include <QDesktopWidget>
#include <QGuiApplication>
#include <QScreen>
#include <QtConcurrent>
#include <QPainter>

#include <QObject>

#include <QDebug>

Screenshot* Screenshot::instance = NULL;

Screenshot* Screenshot::Instance()
{
    if (!instance)
    {
        instance = new Screenshot();
    }
    return instance;
}

Screenshot::Screenshot()
{
    averageColor = new QColor(Qt::white);
    lastAverageColor = new QColor(Qt::white);

    lastRequestedQuality = Settings::Instance()->screenQualityFull;
}

QPixmap *Screenshot::getScreenshot(bool fullscreen, int index)
{
    QPixmap *screenshot;

    screenshot = getScreenshot(index);

    if (fullscreen) Converter::Instance()->scalePixmap(screenshot, Settings::Instance()->screenScaleFull);
    else Converter::Instance()->scalePixmap(screenshot, Settings::Instance()->screenScale);

    if (Settings::Instance()->screenBlackWhite) screenshot = convertToGrayscale(*screenshot);

    return screenshot;
}

QPixmap *Screenshot::getResizedScreenshot(int width)
{
    width = (width > 2000) ? 2000 : width;

    QPixmap *screenshot = getScreenshot(screenIndex);
    if (screenshot->width() > width)
    {
        screenshot = new QPixmap(screenshot->scaledToWidth(width));
    }

    if (Settings::Instance()->screenBlackWhite)
    {
        screenshot = convertToGrayscale(*screenshot);
    }
    return screenshot;
}

QPixmap *Screenshot::getScreenshot(int index)
{
    QList<QPoint*> *locations = getScreenBounds(index);
    QList<QScreen*> screens = QGuiApplication::screens();
    if (screens.length() < index + 1 || index < 0)
    {
        index = 0;
    }

    // When the second screen was right of the primary screen,
    // we had a bug when using the 0th screen as index in screens for taking the screenshot
    // We omit this bug now by using always the last screen and choosing the coordinates plus width/height of the screenshot
    QScreen *screen = QGuiApplication::primaryScreen();
    QPixmap screenshot = screens.at(screens.length() - 1)->grabWindow(0,
                                                       locations->at(0)->x(), locations->at(0)->y(),
                                                       locations->at(1)->x() - locations->at(0)->x(),
                                                       locations->at(1)->y() - locations->at(0)->y());

    return new QPixmap(screenshot);
}

QPixmap *Screenshot::convertToGrayscale(QPixmap &source)
{
    QImage image = source.toImage();
    image = image.convertToFormat(QImage::Format_Mono);
    return new QPixmap(QPixmap::fromImage(image));
}

QList<QPoint *> *Screenshot::getScreenBounds(int index)
{
    QList<QScreen*> screens = QGuiApplication::screens();
    QPoint *startLocation = new QPoint(0, 0);
    QPoint *endLocation = new QPoint(0, 0);

    if (index >= 0)
    {
        QScreen *currentScreen;
        if (index < screens.length()) currentScreen = screens.at(index);
        else
        {
            currentScreen = screens.at(0);
            screenIndex = 0;
        }

        startLocation->setX(currentScreen->geometry().x());
        startLocation->setY(currentScreen->geometry().y());
        endLocation->setX(currentScreen->geometry().x() + currentScreen->geometry().width());
        endLocation->setY(currentScreen->geometry().y() + currentScreen->geometry().height());
    }
    else
    {
        for (int i = 0; i < screens.length(); ++i)
        {
            QScreen *currentScreen = screens.at(i);

            if (currentScreen->geometry().x() < startLocation->x()) startLocation->setX(currentScreen->geometry().x());
            if (currentScreen->geometry().y() < startLocation->y()) startLocation->setY(currentScreen->geometry().y());

            if (currentScreen->geometry().x() + currentScreen->geometry().width() > endLocation->x())
                endLocation->setX(currentScreen->geometry().x() + currentScreen->geometry().width());
            if (currentScreen->geometry().y() + currentScreen->geometry().height() > endLocation->y())
                endLocation->setY(currentScreen->geometry().y() + currentScreen->geometry().height());
        }
    }
    QList<QPoint*> *locations = new QList<QPoint*>();
    locations->append(startLocation);
    locations->append(endLocation);
    return locations;
}

void Screenshot::keepSendingScreenshots(Command &responseCommand)
{
    QtConcurrent::run(this, &Screenshot::keepSendingScreenshotsThread, responseCommand);
}

void Screenshot::keepSendingScreenshotsThread(Command &responseCommand)
{
    isSendingScreenshot = true;
    while (continueSendingScreenshots)
    {
        continueSendingScreenshots = false;
        ApiV3::Instance()->answerScreenGetRequest(responseCommand);
        std::this_thread::sleep_for (std::chrono::milliseconds(screenUpdateInterval));
        ApiV3::Instance()->answerScreenGetRequest(responseCommand);
        std::this_thread::sleep_for (std::chrono::milliseconds(screenUpdateInterval));
        ApiV3::Instance()->answerScreenGetRequest(responseCommand);
    }
    isSendingScreenshot = false;
}

void Screenshot::sendPixmap(QPixmap *pixmap, int quality)
{
    isSendingScreenshot = true;
    Command *command = new Command();
    command->source = Network::Instance()->getServerIp();
    command->destination = Remote::Instance()->lastCommand->source;
    command->priority = Command::PRIORITY_MEDIUM;
    command->data = Converter::Instance()->pixmapToByte(*pixmap, quality);
    command->send();
}

void Screenshot::sendScreenshot(bool full)
{
    if (!isSendingScreenshot)
    {
        if (full) sendPixmap(getScreenshot(true, screenIndex), Settings::Instance()->screenQualityFull);
        else sendPixmap(getScreenshot(false, screenIndex), Settings::Instance()->screenQuality);
    }
}

void Screenshot::toggleScreen()
{
    int new_index = screenIndex + 1;
    if (new_index > QGuiApplication::screens().length() - 1) new_index = -1;
    screenIndex = new_index;
}

void Screenshot::startUpdateColorTimer()
{
    QTimer *timer = new QTimer();
    connect(timer, SIGNAL(timeout()), this, SLOT(updateAverageColorTimerTick()));
    timer->start(colorUpdateInterval);
}

void Screenshot::updateAverageColorTimerTick()
{
    QtConcurrent::run(this, &Screenshot::updateAverageColor);
}

void Screenshot::updateAverageColor()
{
    if (Settings::Instance()->updateAmbientColor)
    {
        averageColor = getApproximateAverageColor(getScreenshot(false, screenIndex));
        if (Settings::Instance()->serialCommands && lastAverageColor->name() == averageColor->name())
        {
            QString message = "<22"
                    + QString::number(Converter::Instance()->byteToAsciiNumber(averageColor->red()))
                    + QString::number(Converter::Instance()->byteToAsciiNumber(averageColor->green()))
                    + QString::number(Converter::Instance()->byteToAsciiNumber(averageColor->blue()))
                    + "9>";
            Serial::Instance()->sendMessage(message);
            lastAverageColor = new QColor(averageColor->name());
        }
    }
}

QColor* Screenshot::getApproximateAverageColor(QPixmap *pixmap)
{
    int totalR = 0;
    int totalG = 0;
    int totalB = 0;
    int count = 1;
    QImage img = pixmap->toImage();
    for (int x = 0; x < pixmap->width(); ++x)
    {
        for (int y = 0; y < pixmap->height(); ++y)
        {
            QColor *pixel = new QColor(img.pixel(x, y));
            if (pixel->saturation() > 0.0 && pixel->value() > 0.0)
            {
                totalR += pixel->red();
                totalG += pixel->green();
                totalB += pixel->blue();
                count += 1;
            }
        }
    }

    int averageR = totalR / count;
    int averageG = totalG / count;
    int averageB = totalB / count;

    int roundUp = 200;
    int roundDown = 100;

    averageR = (averageR > roundUp) ? 255 : ((averageR < roundDown) ? 0 : averageR);
    averageG = (averageG > roundUp) ? 255 : ((averageG < roundDown) ? 0 : averageG);
    averageB = (averageB > roundUp) ? 255 : ((averageB < roundDown) ? 0 : averageB);

    return new QColor(averageR, averageG, averageB);
}
